### PR TITLE
feature/laf-2-oai-update-after-db-change

### PR DESCRIPTION
**Pull request by MajoBerger**
| Phases            | MP | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|----:|-----:|-----:|-------:|
| ETA                  |  0  |  0  |    0 |     0 |      0 |        0 |
| Developing      |  0  |  1  |   8 |    0 |      0 |         0 |
| Review             |  0  |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -  |   -   |   -     |        0 |

## TEST
to test, add, remove or change item. Changes must be visible in OAI and in SOLR

## Problems
- The SolrReindexer throws RuntimeError if something goes wrong during reindexation. It destroyed Integration Tests because SolrOaiService doesn't exist in the DSpace test environment so every reindexation throws RuntimeException. We fixed this problem by removing of throwing Runtime exception and we replaced it by logging.

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+        return authorizeService.authorizeActionBoolean(context, item, Constants.READ);
+    }
+
+
+    private boolean checkIfVisibleInOAI(Item item) throws IOException {
+        SolrQuery params = new SolrQuery("item.id:" + item.getID().toString()).addField("item.public");
+        try {
+            SolrDocumentList documents = DSpaceSolrSearch.query(solrServerResolver.getServer(), params);
+            if (documents.getNumFound() == 1) {
+                return (boolean) documents.get(0).getFieldValue("item.public");
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:106]: What if `documents.get(0)` will be null?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+        // Check if READ access allowed on this Item
+        return authorizeService.authorizeActionBoolean(context, item, Constants.READ);
+    }
+
+
+    private boolean checkIfVisibleInOAI(Item item) throws IOException {
+        SolrQuery params = new SolrQuery("item.id:" + item.getID().toString()).addField("item.public");
+        try {
+            SolrDocumentList documents = DSpaceSolrSearch.query(solrServerResolver.getServer(), params);
+            if (documents.getNumFound() == 1) {
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:105]: Why `== 1` ? Add some comment please.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        } catch (DSpaceSolrException | SolrServerException e) {
+            return false;
+        }
+    }
+
+    private boolean checkIfIndexed(Item item) throws IOException {
+        SolrQuery params = new SolrQuery("item.id:" + item.getID().toString()).addField("item.id");
+        try {
+            SolrDocumentList documents = DSpaceSolrSearch.query(solrServerResolver.getServer(), params);
+            return documents.getNumFound() == 1;
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:119]: Why == 1 ? Add some comment please.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+                return (boolean) documents.get(0).getFieldValue("item.public");
+            } else {
+                return false;
+            }
+        } catch (DSpaceSolrException | SolrServerException e) {
+            return false;
+        }
+    }
+
+    private boolean checkIfIndexed(Item item) throws IOException {
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:115]: Add doc

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+        SolrQuery params = new SolrQuery("item.id:" + item.getID().toString()).addField("item.id");
+        try {
+            SolrDocumentList documents = DSpaceSolrSearch.query(solrServerResolver.getServer(), params);
+            return documents.getNumFound() == 1;
+        } catch (DSpaceSolrException | SolrServerException e) {
+            return false;
+        }
+    }
+
+    private boolean willChangeStatus(Item item) throws SQLException {
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:125]: Add doc

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+        List<ResourcePolicy> policies = authorizeService.getPoliciesActionFilter(context, item, Constants.READ);
+        for (ResourcePolicy policy : policies) {
+            if ((policy.getGroup() != null) && (policy.getGroup().getName().equals("Anonymous"))) {
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:128]: Refactor:
1. Objects.isNull(policy.getGroup()) { continue; }

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+                if (policy.getStartDate() != null && policy.getStartDate().after(new Date())) {
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:129]: Add comment what are you checking here

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+        for (ResourcePolicy policy : policies) {
+            if ((policy.getGroup() != null) && (policy.getGroup().getName().equals("Anonymous"))) {
+                if (policy.getStartDate() != null) {
+                    dates.add(policy.getStartDate());
+                }
+                if (policy.getEndDate() != null) {
+                    dates.add(policy.getEndDate());
+                }
+            }
+            context.uncacheEntity(policy);
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:277]: Refactor

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+        }
+        dates.add(item.getLastModified());
+        Collections.sort(dates);
+        Date now = new Date();
+        Date lastChange = null;
+        for (Date d : dates) {
+            if (d.before(now)) {
+                lastChange = d;
+            }
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:286]: refactor

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java (hunk 1 above)
Comment (MajoBerger) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:106]: all taken from XOAI.java

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java (hunk 2 above)
Comment (MajoBerger) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:105]: all taken from XOAI.java

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java (hunk 3 above)
Comment (MajoBerger) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:119]: all taken from XOAI.java

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java (hunk 4 above)
Comment (MajoBerger) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:115]: all taken from XOAI.java

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java (hunk 5 above)
Comment (MajoBerger) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:125]: all taken from XOAI.java

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java (hunk 6 above)
Comment (MajoBerger) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:128]: all taken from XOAI.java

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java (hunk 7 above)
Comment (MajoBerger) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:129]: all taken from XOAI.java

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java (hunk 8 above)
Comment (MajoBerger) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:277]: all taken from XOAI.java

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java (hunk 9 above)
Comment (MajoBerger) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java:286]: all taken from XOAI.java

**Comment by milanmajchrak** [comment]
@MajoBerger How did you fix the failing tests? - Reindexing of the item in the Solr throwed an error during Integration Tests.

**Review by milanmajchrak — APPROVED** [review]

**Pull request merged**